[Clang][RFC] Bypass TAD during overload resolution if a perfect match exists

cor3ntin opened the pull request:
This implements the same overload resolution behavior as GCC, 
as described in https://wg21.link/p3606 (sections 1-2, not 3)

If, during overload resolution, a non-template candidate is always picked because each argument is a perfect match (i.e., the source and target types are the same), we do not perform deduction for any template candidate that might exist.

The goal is to be able to merge #122423 without being too disruptive.

This change means that the selection of the best viable candidate and template argument deduction become interleaved.

To avoid rewriting half of Clang, we store in `OverloadCandidateSet` enough information to deduce template candidates from `OverloadCandidateSet::BestViableFunction`. This means the lifetime of any object used by the template argument must outlive a call to `Add*Template*Candidate`.

This two-phase resolution is not performed for some initialization as there are cases where template candidates are a better match per the standard. It's also bypassed for code completion.

The change has a nice impact on compile times
https://llvm-compile-time-tracker.com/compare.php?from=719b029c16eeb1035da522fd641dfcc4cee6be74&to=bf7041045c9408490c395230047c5461de72fc39&stat=instructions%3Au .

Fixes #62096
Fixes #74581
Fixes #53454

cor3ntin committed:
[Clang][WIP][RFC] Bypass TAD during overload resolution if a perfect match exists

This implements the same overload resolution behavior as GCC,
as described in https://wg21.link/p3606 (section 1-2, not 3)

If during overload resolution, there is a non-template candidate
that would be always be picked - because each of the argument
is a perfect match (ie the source and target types are the same),
we do not perform deduction for any template candidate
that might exists.

The goal is to be able to merge #122423 without being too disruptive.

This change means that the selection of the best viable candidate and
template argument deduction become interleaved.

To avoid rewriting half of Clang we store in `OverloadCandidateSet`
enough information to be able to deduce template candidates from
`OverloadCandidateSet::BestViableFunction`. Which means
the lifetime of any object used by template argument must outlive
a call to `Add*Template*Candidate`.

This two phase resolution is not performed for some initialization
as there are cases where template candidate are better match
in these cases per the standard. It's also bypassed for code completion.

The change has a nice impact on compile times
https://llvm-compile-time-tracker.com/compare.php?from=719b029c16eeb1035da522fd641dfcc4cee6be74&to=bf7041045c9408490c395230047c5461de72fc39&stat=instructions%3Au

Fixes #62096
Fixes #74581

cor3ntin commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10918,7 +11083,44 @@ bool OverloadCandidate::NotValidBecauseConstraintExprHasError() const {
 OverloadingResult
 OverloadCandidateSet::BestViableFunction(Sema &S, SourceLocation Loc,
                                          iterator &Best) {
+
+  bool TwoPhaseResolution =
+      !NonDeducedCandidates.empty() && Kind != CSK_CodeCompletion &&
+      Kind != CSK_InitByUserDefinedConversion && Kind != CSK_InitByConstructor;
+
+  if (TwoPhaseResolution) {
+    Best = end();
+    for (auto It = begin(); It != end(); ++It) {
+      if (It->isPerfectMatch(S.getASTContext())) {
+        if (Best == end()) {
+          Best = It;
+        } else {
+          Best = end();
+          break;
+        }
+      }
+    }
+    if (Best != end()) {
+      Best->Best = true;
+      if (Best->Function && Best->Function->isDeleted())
+        return OR_Deleted;
+      if (auto *M = dyn_cast_or_null<CXXMethodDecl>(Best->Function);
+          Kind == CSK_AddressOfOverloadSet && M &&
+          M->isImplicitObjectMemberFunction()) {
+        return OR_No_Viable_Function;
+      }
+      return OR_Success;
+    }
+  }
+  InjectNonDeducedTemplateCandidates(S);
+  return BestViableFunctionImpl(S, Loc, Best);
+}
+
+OverloadingResult OverloadCandidateSet::BestViableFunctionImpl(
+    Sema &S, SourceLocation Loc, OverloadCandidateSet::iterator &Best) {
+
```
This is the meat of the change (that and the definition of `isPerfectMatch`)

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10918,7 +11083,44 @@ bool OverloadCandidate::NotValidBecauseConstraintExprHasError() const {
+
+  bool TwoPhaseResolution =
+      !NonDeducedCandidates.empty() && Kind != CSK_CodeCompletion &&
+      Kind != CSK_InitByUserDefinedConversion && Kind != CSK_InitByConstructor;
```
I must be missing something, are `CSK_InitByUserDefinedConversion` and `CSK_InitByConstructor` something special that we can't apply to them?

(explanations are appreciated!)

cor3ntin replied:
I fixed that / added comments

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10918,7 +11083,44 @@ bool OverloadCandidate::NotValidBecauseConstraintExprHasError() const {
+  if (TwoPhaseResolution) {
+    Best = end();
+    for (auto It = begin(); It != end(); ++It) {
+      if (It->isPerfectMatch(S.getASTContext())) {
+        if (Best == end()) {
+          Best = It;
+        } else {
+          Best = end();
+          break;
+        }
+      }
```
If there are two+ perfect matches, I presume the reason for going on adding template candidates is that we want every candidates to be present in the diagnostic?

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10918,7 +11083,44 @@ bool OverloadCandidate::NotValidBecauseConstraintExprHasError() const {
+      Best->Best = true;
+      if (Best->Function && Best->Function->isDeleted())
+        return OR_Deleted;
+      if (auto *M = dyn_cast_or_null<CXXMethodDecl>(Best->Function);
+          Kind == CSK_AddressOfOverloadSet && M &&
+          M->isImplicitObjectMemberFunction()) {
+        return OR_No_Viable_Function;
```
This does not look obvious to me. Can you explain?

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10905,6 +10983,93 @@ bool OverloadCandidate::NotValidBecauseConstraintExprHasError() const {
+static void AddTemplateOverloadCandidate(
+    Sema &S, OverloadCandidateSet &CandidateSet,
+    NonDeducedConversionTemplateOverloadCandidate &&C) {
+  return S.AddTemplateConversionCandidateImmediately(
```
Can we reuse `InjectNonDeducedTemplateCandidates` for `*Immediately` functions or vice versa?

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10975,9 +11176,8 @@ OverloadCandidateSet::BestViableFunction(Sema &S, SourceLocation Loc,
+  llvm::SmallVector<OverloadCandidate *, 4> PendingBest;
   llvm::SmallVector<const NamedDecl *, 4> EquivalentCands;
-
-  llvm::SmallVector<OverloadCandidate*, 4> PendingBest;
```
unrelated changes (sorry for nitpicks at this point)

cor3ntin committed:
avoid comparing types

cor3ntin committed:
Fix logic

cor3ntin committed:
optimize

cor3ntin committed:
optimize again

cor3ntin committed:
Cleanups, fix tests

cor3ntin commented:
I could use help understanding how to fix `clang/test/Templight/templight-empty-entries-fix.cpp` - it always eludes me how to make FileCheck happy @AaronBallman

cor3ntin committed:
Fix templight test

cor3ntin committed:
format

cor3ntin committed:
do not skip templates when initializing by constructors as that might include a conversion

cor3ntin committed:
move and document the special handling of implicit object member function in non-member calls

Sirraide commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -10043,12 +10043,19 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
-      for (Expr *E : Inits)
+
+      // As template candidates are not deduced immediately,
+      // persist the arry in the overload set.
```
```suggestion
      // persist the array in the overload set.
```

cor3ntin replied:
Copy-pasted in a whole bunch of places too!

Sirraide commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -14689,18 +14934,24 @@ void Sema::LookupOverloadedBinOp(OverloadCandidateSet &CandidateSet,
   AddNonMemberOperatorCandidates(Fns, Args, CandidateSet);
 
+  // As template candidates are not deduced immediately,
+  // persist the arry in the overload set.
```
```suggestion
  // persist the array in the overload set.
```

Sirraide reviewed (commented):
Just a few typos I noticed

zygoloid commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10230,9 +10307,15 @@ Sema::AddArgumentDependentLookupCandidates(DeclarationName Name,
               *this, Args, FTD->getTemplatedDecl())) {
+
+        // As template candidates are not deduced immediately,
+        // persist the arry in the overload set.
```
```suggestion
        // persist the array in the overload set.
```

zygoloid reviewed (commented):
I think we should do this. Even if WG21 decides that they only want to allow this and not require it, it seems like a good change that we stand a decent chance of reaching implementation consensus on.

cor3ntin committed:
Fix typos and cuda tests

cor3ntin committed:
remove redundant check

zyn0217 commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1126,7 +1208,8 @@ class Sema;
     OperatorRewriteInfo RewriteInfo;
 
     constexpr static unsigned NumInlineBytes =
-        24 * sizeof(ImplicitConversionSequence);
+        32 * sizeof(ImplicitConversionSequence);
```
Can we add some comments? what causes us to change it from 24 -> 32

cor3ntin replied:
I added some. We may be storing additional data in the slab allocator ( When we need to keep the argument array expression around for deferred candidates), so I made the storage a tad bigger to accomodate.  There is no reason for this thing to not be quite large frankly. it's short lived

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -10933,27 +11096,100 @@ OverloadCandidateSet::BestViableFunction(Sema &S, SourceLocation Loc,
+
+  Best = end();
+  for (auto It = begin(); It != end(); ++It) {
+    if (It->isPerfectMatch(S.getASTContext())) {
```
maybe
```cpp
if (!It->isPerfectMatch(S.getASTContext()))
  continue;
```
to reduce the curley nesting.

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -14354,6 +14584,17 @@ ExprResult Sema::BuildOverloadedCallExpr(Scope *S, Expr *Fn,
   OverloadingResult OverloadResult =
       CandidateSet.BestViableFunction(*this, Fn->getBeginLoc(), Best);
 
+  // [C++23][over.call.func]
+  // if overload resolution selects a non-static member function,
+  // the call is ill-formed;
+  if (CSK == OverloadCandidateSet::CSK_AddressOfOverloadSet &&
+      Best != CandidateSet.end()) {
```
Hmmm. I'm surprised this wasn't caught until now. Do you have a test case? (or is this something copied from elsewhere?)

zyn0217 replied:
nevermind I realized it is https://github.com/llvm/llvm-project/pull/133426#discussion_r2018513280 :P

zyn0217 reviewed (commented):
Thanks!

cor3ntin committed:
More code simplification

cor3ntin committed:
fix cuda, add comments

cor3ntin committed:
add tests

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1116,17 +1197,22 @@ class Sema;
   private:
     SmallVector<OverloadCandidate, 16> Candidates;
     llvm::SmallPtrSet<uintptr_t, 16> Functions;
+    SmallVector<DeferredTemplateOverloadCandidate, 8> DeferredCandidates;
```
This is DEFINITELY cleaner, but a future 'optimization' here might store a single candidates list, then maintain a single 'index' to the first deferred candidate.  This would save a potential allocation/potentially improve perf, but at the expense of some complexity.  

I AM somewhat concerned about the increased size here though.  That SmallVector here is going to be HUGE, and Candidates is already quite large.  These are done often on the stack unfortunately, so this is going to have some effect on instantiation depth.

cor3ntin replied:
I actually thought about that - but I have not implemented it yet.
Are the cases where we do a lot of overload resolution recursively that common?

erichkeane replied:
I recall Aaron I thought at one point working on reducing the overload-candidates size which showed great improvement, but that would have been ~9 months ago now.  I remember being surprised at how much of a difference it made.

erichkeane reviewed (commented):
I agree completely with Richard, we should just do this.  We have prior art (GCC), the standard currently permits it as an implementation strategy, and it is an improvement in basically EVERY way with mild/barely-justifiable downsides.

cor3ntin committed:
* Fix handling of explicit specifiers
* Prefer a constructor template over a non-template conversion function

cor3ntin committed:
slab allocate template candidates

cor3ntin committed:
fix rvalue to lvalue binding

cor3ntin committed:
Disable resolution by perfect match when
 - We find a template conversion candidate
 - A constructor with a dependent explicit specifier

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -407,6 +407,18 @@ class Sema;
              Third == ICK_Identity;
     }
 
+    bool isPerfect(const ASTContext &C) const {
```
I would LOVE a comment here, the implementation isn't particularly clear what is going on here.

erichkeane replied:
Still would like comments on these functions that say what 'perfect' means here.

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -979,6 +995,20 @@ class Sema;
       return false;
     }
 
+    bool isPerfectMatch(const ASTContext &Ctx) const {
```
`isPerfect` vs `isPerfectMatch` are... confusing.  Work to make these easier to understand together would be appreciated.

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1117,16 +1205,27 @@ class Sema;
 
-    // Allocator for ConversionSequenceLists. We store the first few of these
+    DeferredTemplateOverloadCandidate *FirstDeferredCandidate;
+    unsigned DeferredCandidatesCount : 8 * sizeof(unsigned) - 2;
```
isn't this size possibly different on platforms, and thus makes the `32` on 1227 unreliable?

cor3ntin replied:
The 32 on line 1227 is super arbitrary and not tied to the size of `unsigned`.
It did felt worth making bigger as we are now storing more than conversion sequences in the overload set.

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1015,6 +1045,62 @@ class Sema;
   };
 
+  struct DeferredTemplateOverloadCandidate {
+    DeferredTemplateOverloadCandidate *Next = nullptr;
```
Comment here + on the slab allocator function as to WHY we are doing this instead of a std::vector/etc would probably be helpful.  I understand it, but I probably won't next time I see this code ;)

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1159,12 +1256,29 @@ class Sema;
     }
 
+    template <typename T> T *allocateDeferredCandidate() {
+      T *C = slabAllocate<T>(1);
```
Seems funny to me to `slab` allocate in increments of 1.  I have a 'silly' idea... Since 'deferred candidates' are likely to only be a handful per lookup... I wonder if we could:

1- Add a few-bits to `DeferredTemplateOverloadCandidate` for `leftInSlab`.  
2- on `!FirstDeferredCandidate`, allocate (say)4 items.  Set the 1st item to '3' left.
3- Then, find the last candidate the same way.  if `leftInSlab` != 0, emplacement-new/etc into the 'next' location, set the `leftInSlab` to be the previous `-1`.  If `leftInSlab == 0`, allocate.

I realize it is more work here, and can be done in a followup, but there might be a mild perf improvement for cache locality.  I suspect if we set this to something like 4, which requires 2 bits in the structure, high-90%s of lookups will take only 1 allocation.

WDYT?

erichkeane replied:
Thinking further, the whole point of the slab is to put all the stuff in the same location, so this is a silly suggestion.  Disregard.

erichkeane reviewed (commented):
A few mean comments, but the approach still seems solid here.  I think this is a solid patch

cor3ntin committed:
address Erich's, format, add release notes

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -743,6 +759,13 @@ class Sema;
     }
 
+    /// A conversion sequence is perfect if
+    /// it is an identity conversion and
```
your formatting of this comment is a little weird re-newlines.

erichkeane commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -10382,6 +10383,7 @@ class Sema final : public SemaBase {
   /// Add a C++ function template specialization as a candidate
   /// in the candidate set, using template argument deduction to produce
   /// an appropriate function template specialization.
+
```
Pretty sure we don't want this :)

erichkeane reviewed (approved):
I have a handful of comments that are all nits/minor enough I trust you on, but overall this looks fine. 

Lets ship it!

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1043,6 +1141,8 @@ class Sema;
       /// Resolve a call through the address of an overload set.
       CSK_AddressOfOverloadSet,
+
+      CSK_CodeCompletion,
```
Ah, I see.  Can you clarify that in a comment above this as to why this is a separate category?  It was pretty jarring to see in general, let alone here.

cor3ntin committed:
address more of Erich's feedback

cor3ntin committed:
Clarify changelog

cor3ntin committed:
fix assertion

erichkeane commented on clang/include/clang/Sema/Overload.h:
```diff
@@ -1216,6 +1352,24 @@ class Sema;
+    llvm::MutableArrayRef<Expr *> getPersistentArgsArray(T *...Exprs) {
+      llvm::MutableArrayRef<Expr *> Arr =
+          getPersistentArgsArray(sizeof...(Exprs));
+      llvm::copy(std::initializer_list<Expr *>{Exprs...}, Arr.data());
```
ah, hrmph, thats unfortunate.  There is _A_ good reason that was brought up on me in the past, but everything I can remember/think of is sort of a "well, i guess it is 'more' correct'" when it comes to pointers. So don't change it here, as the init-list stuff makes it more error prone.

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -7919,7 +7951,7 @@ void Sema::AddTemplateOverloadCandidate(
     Candidate.Viable = false;
     Candidate.RewriteKind =
-      CandidateSet.getRewriteInfo().getRewriteKind(Candidate.Function, PO);
+        CandidateSet.getRewriteInfo().getRewriteKind(Candidate.Function, PO);
```
unrelated?

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -7834,7 +7838,7 @@ void Sema::AddMethodTemplateCandidate(
     Candidate.Viable = false;
     Candidate.RewriteKind =
-      CandidateSet.getRewriteInfo().getRewriteKind(Candidate.Function, PO);
+        CandidateSet.getRewriteInfo().getRewriteKind(Candidate.Function, PO);
```
ditto

zyn0217 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -45,6 +45,7 @@
 #include <cstddef>
 #include <cstdlib>
 #include <optional>
+#include <variant>
```
We no longer use variant, right?

zyn0217 reviewed (approved):
Some nits, and thanks for helping us in this area!

cor3ntin committed:
cleanups

nikic commented:
The final version had a 4% improvement on clang bootstrap: https://llvm-compile-time-tracker.com/compare.php?from=23324b8b109aed1f77cb20cef476b795f33b6835&to=8c5a307bd8d406e6167a5cd3ce3c74e2e3bfb2a6&stat=instructions:u

Wow.
